File: engine/src/main/java/model/moves/EnPassantMove.java
package model.moves;

import model.GameData;
import model.pieces.Pawn;
import model.pieces.Piece;
import model.util.Position;
import org.jetbrains.annotations.NotNull;

public class EnPassantMove extends BaseMove {
    private Piece eatenPiece;

    public EnPassantMove(@NotNull Position start, @NotNull Position destination) {
        super(start, destination);
    }

    @Override
    void applyToGame(@NotNull GameData data) {
        super.applyToGame(data);

        eatenPiece = data.getBoard().removePiece(end.shift(((Pawn) piece).BACKWARD));
        data.getEatenPieces(eatenPiece.getColour()).push(eatenPiece);
    }

    @Override
    void undoToGame(@NotNull GameData data) {
        data.getEatenPieces(eatenPiece.getColour()).pop();
        data.getBoard().add(end.shift(((Pawn) piece).BACKWARD), eatenPiece);

        super.undoToGame(data);
    }

    @Override
    public int getValue() {
        return -eatenPiece.getSignedValue();
    }
}

File: engine/src/main/java/model/moves/EnPassantMove.kt
package model.moves

import model.GameData
import model.pieces.Pawn
import model.pieces.Piece
import model.util.Position

internal class EnPassantMove(start: Position, destination: Position) : BaseMove(start, destination) {
    private var eatenPiece: Piece? = null

    override val value: Int
        get() = -eatenPiece!!.signedValue

    override fun applyToGame(data: GameData) {
        super.applyToGame(data)

        eatenPiece = data.board.removePiece(end.shift((piece as Pawn).backward))
        data.getEatenPieces(eatenPiece!!.colour).push(eatenPiece)
    }

    override fun undoToGame(data: GameData) {
        data.getEatenPieces(eatenPiece!!.colour).pop()
        data.board.add(end.shift((piece as Pawn).backward), eatenPiece!!)

        super.undoToGame(data)
    }
}
